fallback.go

```diff
@@ -0,0 +1,144 @@
+package txtdirect
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+)
+
+type Fallback struct {
+	rw      http.ResponseWriter
+	request *http.Request
+	config  Config
+
+	records    []record
+	pathRecord record
+	lastRecord record
+
+	fallbackType string
+	code         int
+}
+
+// fallback redirects the request to the given fallback address
+// and if it's not provided it will check txtdirect config for
+// default fallback address
+func fallback(w http.ResponseWriter, r *http.Request, fallbackType string, code int, c Config) {
+	if code == http.StatusMovedPermanently {
+		w.Header().Add("Cache-Control", fmt.Sprintf("max-age=%d", status301CacheAge))
+	}
+	w.Header().Add("Status-Code", strconv.Itoa(code))
+
+	f := Fallback{
+		rw:           w,
+		request:      r,
+		config:       c,
+		fallbackType: fallbackType,
+		code:         code,
+	}
+
+	if fallbackType != "global" {
+		// Fetch records from request's context and set the []record type on them
+		f.fetchRecords()
+
+		// Redirect to first record's `to=` field
+		if (fallbackType == "to" && f.lastRecord.To != "") && f.validateURIs() {
+			http.Redirect(w, r, f.lastRecord.To, code)
+			f.countFallback(f.lastRecord.Type)
+		}
+
+		// Redirect to first record's `website=` field
+		if fallbackType == "website" && f.lastRecord.Website != "" {
+			http.Redirect(w, r, f.lastRecord.Website, code)
+			f.countFallback(f.lastRecord.Type)
+		}
+
+		// Dockerv2 root fallback
+		if fallbackType == "root" && f.lastRecord.Root != "" {
+			http.Redirect(w, r, f.lastRecord.Root, code)
+			f.countFallback(f.lastRecord.Type)
+		}
+
+		if fallbackType == "root" && f.pathRecord.Root != "" {
+			http.Redirect(w, r, f.pathRecord.Root, code)
+			f.countFallback(f.pathRecord.Type)
+		}
+
+		if f.pathRecord.To != "" {
+			http.Redirect(w, r, f.pathRecord.To, code)
+			f.countFallback(f.pathRecord.Type)
+		}
+
+		// If non of the above cases applied on the record, jump into global redirects
+		f.globalFallbacks(f.lastRecord.Type)
+		log.Printf("[txtdirect]: %s > %s", r.Host+r.URL.Path, w.Header().Get("Location"))
+		return
+	}
+
+	f.globalFallbacks("")
+
+	log.Printf("[txtdirect]: %s > %s", r.Host+r.URL.Path, w.Header().Get("Location"))
+}
+
+func addRecordToContext(r *http.Request, rec record) *http.Request {
```

Comment:
We can make it a bit more customizable and add another parameter like `contextKey` and declare a constant variable inside fallback.go which contains the `fallback_records` value. That way we can store records inside the context with different rules applied for each use case and separated by the contextKey.

Not sure it would be useful or not, but anyway I agree that it fits better inside the record.go file.

Reply:
As it's just the general record object being stored no additional contextKey is necessary for now.
Basically moving it to record.go should be sufficient and fallback just uses the general record context then.